### ISSUE TITLE
OptionParser.to_argv/1 cannot deal with :count switches

### ISSUE THREAD

**Issue by ThomasArts** [elixir-lang/elixir#5327]
### PreCheck

This is between a bug report and a feature request...
### Environment
- Elixir version (elixir -v): Elixir 1.4.0-dev (c07c345) on Erlang R19
- Operating system: MacOS
### Current behavior

Assume you have a list of options in which of them is a `:count`, i.e., can occur several times and only the number of times it occurs is taken into account.

The issue I have is translating this back to an argv list again.

``` elixir
  test "Counts cannot be translated back" do
    original = ["--count1", "--count1"]
    {opts, [], []} = OptionParser.parse(original,  [switches: @switches])
    assert original == OptionParser.to_argv(opts)
  end
```

Sine `:count1` is specified as `:count` the opts contain `count: 2` after parsing. But there is no way for the reverse operation to_argv to know that this was a count option. Executing the test therefore fails:

```
  1) test Counts cannot be translated back (OptionParserEqc)
     test/option_parser_eqc.exs:79
     Assertion with == failed
     code:  original == OptionParser.to_argv(opts)
     left:  ["--count1", "--count1"]
     right: ["--count1", "2"]
     stacktrace:
       test/option_parser_eqc.exs:82: (test)
```

Another asymmetry that I found slightly disturbing is that arguments are treated different from options in the to_argv translation. Probably there is a good reason for it, but my first approach was to give to_argv what I got back from the parser:

``` elixir
    {opts, args, []} = OptionParser.parse(original,  [switches: @switches])
    assert original == OptionParser.to_argv(opts ++ args)
```

This could have worked, but fails badly: with an error message that isn't obvious.

``` elixir
 test "No real symmetry in operations" do
    argv = ["--b1", "--no-b2", "filename"]
    {opts, args, []} = OptionParser.parse(argv,  [switches: @switches])
    assert argv == OptionParser.to_argv(opts ++ args)
  end
```

```
 2) test No real symmetry in operations (OptionParserEqc)
     test/option_parser_eqc.exs:85
     ** (FunctionClauseError) no function clause matching in anonymous fn/1 in OptionParser.to_argv/1
     stacktrace:
       (elixir) lib/option_parser.ex:399: anonymous fn("filename") in OptionParser.to_argv/1
       (elixir) lib/enum.ex:966: Enum.flat_map_list/2
       (elixir) lib/enum.ex:967: Enum.flat_map_list/2
       test/option_parser_eqc.exs:90: (test)



```
### Proposed behavior

To remedy the above one could supply a version of to_argv that also gets the switches as argument.

``` elixir
 def to_argv(elements, options) do
    switches = options[:switches]
    {opts, args} =
      Enum.reduce(elements, {[], []},
        fn({k,v}, {os, as}) ->
          if {k, :count} in switches do
            { os ++ List.duplicate({k, true}, v), as}
          else
            {os ++ [{k,v}], as}
          end
          (arg, {os, as}) ->
            {os, as ++ [arg]}
        end)
    OptionParser.to_argv(opts) ++ args
  end
```

This is backward compatible, but behaves symmertic.

I have tested the above code with QuickCheck for any possible option and argument list.
The QuickCheck specification is:

``` elixir
 property "OptionParser.to_argv as I like to have it" do
    forall {opts, arguments} <- {options(), args()} do
      argv = to_argv(opts ++ arguments, [switches: @switches])
      when_fail IO.puts "Argv #{inspect argv}" do
        {new_opts, new_arguments, errors} = OptionParser.parse(argv, [switches: @switches])
        conjunction(errors: (ensure errors == []),
                    extra_arguments: (ensure new_arguments -- arguments == []),
                    missing_arguments: (ensure arguments -- new_arguments == []),
                    extra_options: (ensure new_opts -- opts == []),
                    missing_options: (ensure opts -- new_opts == []))
      end      
    end
  end
```

With generators defined as below. Among others, I found out that the strings may not start with a "-", of course. 

``` elixir
@switches [b1: :boolean, b2: :boolean, count1: :count, count2: :count,
             b3: [:boolean, :keep], b4: [:keep, :boolean],
             int1: :integer, int2: :integer, int3: [:integer, :keep], int4: [:keep, :integer],
             f1: :float, f2: :float, f3: [:float, :keep], f4: [:keep, :float],
             str1: :string, str2: :string, str3: [:string, :keep], str4: [:keep, :string],
             str5: :keep
            ]

  defp pos do
    let n <- nat(), do: n+1
  end

  defp option({key, type}) do
      case type do
        :boolean   -> {key, bool()}
        :count     -> {key, pos()}
        [:keep, t] -> option({key, t})
        [t, :keep] -> option({key, t})
        :keep      -> option({key, :string})
        :string    -> {key, string()}
        :float     -> {key, real()}
        :integer   -> {key, int()}
      end
  end

  defp keep(t) do (t == :keep) || (is_list(t) && :keep in t) end

  defp options do
    let {base_opts, keep_opts} <- {sublist(Enum.filter(@switches, fn({k,t}) -> not keep(t) end)),
                                    list(elements(Enum.filter(@switches, fn({k,t}) -> keep(t) end)))} do
      let opts<- shuffle(base_opts ++ keep_opts) do
        for opt<-opts, do: option(opt)
      end
    end
  end

  defp string do
    such_that str <- non_empty(utf8()), do: String.first(str) != "-"
  end

  defp args do
    list(string())
  end
```

**Comment by josevalim** [comment]
Excellent bug report! Thank you @ThomasArts.

Quick question. Shouldn't the real symmetry test be written as:

```elixir
 test "No real symmetry in operations" do
    argv = ["--b1", "--no-b2", "filename"]
    {opts, args, []} = OptionParser.parse(argv,  [switches: @switches])
    assert argv == OptionParser.to_argv(opts) ++ args
  end
```

i.e. we can only convert the options back to argv, the args should be concatenated after.

In any case, we will likely need the switches version as well as improve the error reporting side.

**Comment by ThomasArts** [comment]
I wrote the test like this on purpose. If you are not the implementer, but user of tf the library, it appears to be the the following types apply:

```
OptionParser.parse(argv()) :: {opts(), args(), errors()}

OptionParser.to_argv(list()) :: argv()
```

In other words, parse gives you something from an argv and the inverse is that you give that something back to to_argv and get your original argv back.

Imagine that you change the type of argv() to a %Argv struct instead of a list. Then code that used to_argv(...) ++ something will break. In other words, the type has not nicely been encapsulated in the function abstraction.
In order to have completely opaque types, one actually should either specify that opts and args are lists, or keep them abstract and make a 3 argument to_argv:
`to_argv(opts, args, options())` where options contains the `switches: switches`

In summary: make the functions in such a way that the user of the library need not know the underlying type implementation.

**Comment by muaazsaleem** [comment]
Pushed a first draft for this issue. 

Since both `--count1 2` and `--count1 --count1` compile to `count: 2`,  passing in an optional `opts` argument could solve the issue.

If you guys agree with this approach I could add the appropriate documentation.
